src/modules/hip/kernel/tensor_mean.hpp

```diff
@@ -0,0 +1,483 @@
+    partialSum_smem[hipThreadIdx_x] += (src_f8.f1[0] +
+                                        src_f8.f1[1] +
+                                        src_f8.f1[2] +
+                                        src_f8.f1[3]);              // perform small work of reducing float4s to float using 1024 x 1 threads and store in Shared
+    __syncthreads();                                                // syncthreads after Shared load
+
+    // Reduction of 1024 floats on 1024 threads per block in x dimension
+    for (int threadMax = 512; threadMax >= 1; threadMax /= 2)
+    {
+        if (hipThreadIdx_x < threadMax)
+            partialSum_smem[hipThreadIdx_x] += partialSum_smem[hipThreadIdx_x + threadMax];
+        __syncthreads();
+    }
+
+    // Final store to dst
+    if (hipThreadIdx_x == 0)
+    {
+        int totalElements = roiTensorPtrSrc[hipBlockIdx_z].xywhROI.roiHeight * roiTensorPtrSrc[hipBlockIdx_z].xywhROI.roiWidth;
+        dstPtr[hipBlockIdx_z] = partialSum_smem[0] / totalElements;
+    }
+}
+
+__global__ void tensor_mean_grid_3channel_result_hip(Rpp32u *srcPtr,
+                                                     uint xBufferLength,
+                                                     float *dstPtr,
+                                                     RpptROIPtr roiTensorPtrSrc)
+{
+    int id_x = hipThreadIdx_x * 8;
+    int id_z = hipBlockIdx_z;
+
+    __shared__ uint partialRSum_smem[1024];                                      // 8192 uints of src reduced to 1024 in a 1024 x 1 thread block
+    __shared__ uint partialGSum_smem[1024];
+    __shared__ uint partialBSum_smem[1024];
+    partialRSum_smem[hipThreadIdx_x] = 0;                                        // initialization of Shared to 0 using all 1024 x 1 threads
+    partialGSum_smem[hipThreadIdx_x] = 0;
+    partialBSum_smem[hipThreadIdx_x] = 0;
+
+    if (id_x >= xBufferLength)
+        return;
+
+    int xAlignedLength = xBufferLength & ~7;                                     // alignedLength for vectorized global loads
+    int xDiff = xBufferLength - xAlignedLength;                                  // difference between bufferLength and alignedLength
+    uint srcIdx = ((id_z * xBufferLength) + id_x) * 3;
+
+    d_uint24 src_ui24;
+    *reinterpret_cast<d_uint24_s *>(&src_ui24) = *reinterpret_cast<d_uint24_s *>(srcPtr + srcIdx);
+    rpp_hip_layouttoggle24_pkd3_to_pln3(reinterpret_cast<d_uint24_s *>(&src_ui24));
+
+    if (id_x + 8 > xBufferLength)                                                // local memory reset of invalid values (from the vectorized global load) to 0.0f
+    {
+        for(int i = xDiff; i < 8; i++)
+        {
+            src_ui24.ui8[0].ui1[i] = 0;
+            src_ui24.ui8[1].ui1[i] = 0;
+            src_ui24.ui8[2].ui1[i] = 0;
+        }
+    }
+    src_ui24.ui8[0].ui4[0] += src_ui24.ui8[0].ui4[1];
+    src_ui24.ui8[1].ui4[0] += src_ui24.ui8[1].ui4[1];
+    src_ui24.ui8[2].ui4[0] += src_ui24.ui8[2].ui4[1];
```

Comment:
Regarding the changes for mean and https://github.com/r-abishek/rpp/pull/143/files#r1568315323:

Just for this function, my suggestion is to club all Rpp8u/8s/16/32f - all 4 in 1 for tensor_mean_grid_result_hip() and again all 4 in 1 for tensor_mean_grid_3channel_result_hip()
You could do this by using d_uint24_s or d_float24_s that use a very generic data[] array inside the struct. So the struct-members are type independent as they are all *.data[], and you won't have to address them as *.ui[] or *.i[] or *.f[]
So, templating would be easier.

The only issue would be in L185-187 with vector-4 adds. Try typecasted vector-4 adds, else check performance on a modular, templated helper to do the 12 adds independently.

Reply:
Looks much more clean now